### PR TITLE
Allow `any` and `all` only for all-`NA` and empty string columns

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description
This PR allows `any` and `all` for all-NA string columns and string columns that have size 0. This is an essential workaround for time-being because `any` and `all` aren't natively supported for `string` types in `libcudf` and without these workarounds, multiple places in the reduction APIs will need if/elif checks which will make it harder to maintain.

This PR:
```
= 5 failed, 102252 passed, 2090 skipped, 976 xfailed, 312 xpassed in 1375.59s (0:22:55) =
```
On `pandas_2.0_feature_branch`:
```
= 9 failed, 102247 passed, 2091 skipped, 976 xfailed, 312 xpassed in 1336.47s (0:22:16) =
```

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by galipremsagar** [comment]
I think long-term we can get `any` and `all` implemented in python layer for string column with the existing set of string size calculation APIs.

File: python/cudf/cudf/core/column/string.py [hunk 1]
```diff
@@ -5594,16 +5594,15 @@ def data(self):
         return self._data
 
     def all(self, skipna: bool = True) -> bool:
-        # The skipna argument is only used for numerical columns.
-        # If all entries are null the result is True, including when the column
-        # is empty.
-
+        if self.null_count == self.size:
+            return True
```
Comment (bdice) [inline python/cudf/cudf/core/column/string.py:5598]: If `skipna` is `False`, pandas 2.1.4 raises an error. Do we need to check that here?

```python
import pandas as pd
s = pd.Series([pd.NA, pd.NA])
print(s.all(skipna=True))  # True
print(s.all(skipna=False))  # TypeError: boolean value of NA is ambiguous
```

```suggestion
        if skipna and self.null_count == self.size:
            return True
```

File: python/cudf/cudf/core/column/string.py [hunk 2]
```diff
@@ -5594,16 +5594,15 @@ def data(self):
...
         raise NotImplementedError("`all` not implemented for `StringColumn`")
 
     def any(self, skipna: bool = True) -> bool:
-        # The skipna argument is only used for numerical columns.
-        # If all entries are null the result is True, including when the column
-        # is empty.
+        if not skipna and self.has_nulls():
+            return True
```
Comment (bdice) [inline python/cudf/cudf/core/column/string.py:5603]: If `skipna` is `False` and there are any `pd.NA` values, pandas 2.1.4 raises an error. This code would return `True`, from what I can tell. Do we need to raise an error?

```python
import pandas as pd
s = pd.Series([pd.NA, pd.NA])
print(s.any(skipna=True))  # False
print(s.any(skipna=False))  # TypeError: boolean value of NA is ambiguous
```

Comment (wence-) [reply python/cudf/cudf/core/column/string.py:5603]: Yes please!

File: python/cudf/cudf/core/column/string.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/string.py]: Thanks for catching this @bdice, fixed it.

File: python/cudf/cudf/core/column/string.py (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/string.py]: Fixed it.

**Review by bdice — APPROVED** [review]

**Pull request merged**